### PR TITLE
CreateAccount - New Use Case for Creating an Account

### PR DESCRIPTION
**Pull request by JoshJ24**
This pull request features the new use case CreateAccount which allows the user to create an account and adds this account to their list of accounts which is also updated accordingly in the table. 

This pull request also features many bug fixes and changes made within the dashboard use case, specifically in the dashboard view as the dashboard view and createAccount view are linked. The significant changes in this pull request include the following:

- addAccount method was implemented in DAO where given strings of user inputs, creates an account through AccountFactory, then adds the account to the existing list of accounts and stores in database
- updated encrypt and decrypt functions in DAO to account for when empty strings are entered by adding a whitespace character for an empty string
- The views for the buttons on dashboard view such as Create(corresponding to createAccount) will take up the right panel in the Dashboard view, thus a rightPanelView was added as a variable in the Dashboard state along with setter and getter methods in order to keep track of which view should be displayed in the right panel according to the appropriate user interaction 

DashboardView Changes:
-separate views are created for each use case, however for those views that are to be displayed in the right panel of dashboard view, a panel assigned to the particular use case's view is initialized in the constructor for dashboard view 
-for example in this use case, there is a createAccountPanel that is assigned the CreateAccountView, with the  appropriate parameters (view model, etc.) passed through as well
-In the action listeners for each button, the action performed method removes the right panel from main and adds the 
appropriate use case panel in its place. Thus in this case, within the action performed method of the createButton action 
listener, the createAccountPanel is added to main to replace the right panel
 -in the same action performed method, a helper method is called that sets the right panel name to the name of the 
corresponding button/use case, and the variable RightPanelView stated above in the dashboard state is passed that name
-Finally in the propertyChange method for dashboard view, an if statement is added to first check if the rightPanelView defined in the dashboard state is not null, and then checks if it is equal to "dashboard"(for example if the cancel button was pressed we want to go to the original dashboard view), in which case it updates main to add the original rightPanel from dashboard view
-Still in the propertyChange method, a while loop is then added after the above if statements to remove any existing rows from the table so that the table does not retain data from previous accounts, and finally the table is filled with the appropriate list of accounts corresponding to the desired user

In summary, these changes allow smooth insertions of different views into the right panel of dashboard view, while also allowing the smooth transition between these views by consistently keeping track of the desired RightPanelView.

### CODE REVIEW

**Comment by hadi-naqvi** [comment]
**Coding Style:**
The spacing/indentation conventions of all the newly implemented code is consistent with the rest of the codebase. Additionally, the appropriate selection between public and private methods/attributes is consistent with the design choices made in other parts of the program. Furthermore, the variable naming conventions are also consistent with the other parts of the program, which includes the naming conventions of JComponents in our Swing UIs which was previously seen to be inconsistent in others parts of our program such as the SetupAuth and Authentication views. While those views need their variable names to be changed to be consistent, the variable names for this use case do not need to change, as they already match the majority of our codebase's variable naming conventions. 

**Correctness:**
The purpose of this use case is to allow the user to securely store new login data in their password manager. This login data includes the title of their login data, username, password, 2FA secret key, url, iconURL (the icon of the service/website the account is for), date, and other notes related to the login information. This use case allows the user to do exactly that, while accounting for any potential errors. In the success case, a new account is created and added to the database, and the dashboard is updated to reflect the changes (i.e. adds the account to the JTable of accounts). Likewise in the fail case, if the user already has an account with the same title and username, it shows a popup indicating that they are attempting to create a duplicate account and the account is not created. Moreover, the implementaiton of this use case requires security, and it correctly encrypts the user's login data before storing it in the database correctly, and also empties the textfields and updates the state to be empty in the create account UI after account creation is complete as needed. This is important for correctness of security, because otherwise sensitive data (the login details) could potentially be accessed by foreign users.

This pull request also modifies the DisplayDash use case which is responsible for updating the dashboard's table of accounts with the most up to date list of accounts. It was modified so that the table is first emptied, and then the accounts are added back in to reflect the change. Previously the table wasn't emptied and so there would have been duplicate accounts listed in the table. Therefore since it was modified as required, this part of the use case's implementation is also correct and works as needed.

**Presence of tests and Test coverage:**
This pull request features no new tests or test coverage, however this is fine as test coverage is expected to be implemented shortly in another branch/pull request. As of right now, the use case's expected functionality has sufficiently been tested and shown to be correct, so this pull request can be merged into the main branch.

**Adherence to Clean Architecture + SOLID:**
All the application and business rules are segregated as needed. This use case follows the chain of going from the view, to the controller, to the use case interactor, to the presenter, which updates the view model, which then updates the view. This chain of events in which the use case occurs follows the Clean Architecture engine and design patterns that the rest of the use cases in the program have been using, and is not only consistent, but correctly applies Clean Architecture. Additionally, the dependency-inversion principle is not broken in this use case as all necessary components such as DAOs are being injected from the outer layers into the inner layers which require them. For example, in the app package which contains the use case factories, the interactor is instantiated with the DAO since it requires it, however the interactor itself does not have an inverse dependency on any of the outer layers. There is also a DataAccessInterface for this use case which dictates which methods the DAO must implement for the use case.

The SOLID principles in this use case have not been violated and have been followed well. A new design choice that was made in this pull request was to create separate views/forms using the Swing UI Designer in IntelliJ for the views which are going to be displayed in the right panel of the dashboard. Previously, our program was going to simply have all the views in the dashboard view. If we had done so, it would have been a violation of the single responsibility principle since the dashboard view would be responsible for multiple different views, and would have more than one purpose/reason to change which could introduce bugs in the future. As such, the current approach is ideal and follows the single responsibility principle well as it segregates different responsibilities/views into different files, and then instantiates the different views inside the dashboard view as needed.

**Javadocs:**
While most public methods are documented well with javadocs, there are some methods which are left without a Javadoc still, such as the switchView() method which includes the interface CreateAccountOutputBoundary, the controller, the use case interactor, and presenter for the use case (all of which have the method). Otherwise, all other methods have sufficient and well detailed documentation describing method purposes, parameters, and return types.

**Clarification of any accompanying information in the pull request:**
The accompanying information in this pull request touches upon most of the bug fixes, design changes, and new additions in this pull request. However there are some changes which are incorrectly explained or could be better explained for other readers and I will clarify them below:
- The encrypt() and decrypt() methods in the DAO responsible for encrypting and decrypting our login data were updating to handle the edge case of the data being encrypted/decrypted being an empty string (i.e. when the user creates an account, they do not fill in one of the fields, which is something they are allowed to do as not all fields are necessarily required and that part is left up to the user). The methods now check if the data being encrypted/decrypyted is an empty string, and simply return an empty string back as nothing else needs to be done. Previously, it would attempt to run the encryption/decryption algorithm and result in errors/not work correctly. Note that although the pull request mentions whitespaces, there are no whitespaces in our program anywhere, just empty strings.
- In the future, whenever the right panel view is changed to display another view, its rightPanelView name attribute must be updated accordingly (initially on program startup it is null which is fine). When exiting from the view of another use case with a cancel/close button, you will need to update that attribute name with "dashboard" and then trigger "dashboardViewModel.firePropertyChanged()" in order to switch the right panel view back to the original dashboard. For further information, take a look into the code to understand how this process works and the design choices made while implementing the logic for this.
- The DashboardView file now has helper methods for operations requiring the "this" keyword inside of action listeners. This was necessary because inside the frame/context of an action listener, "this" refers to the ActionListener object and not the view itself. Therefore to referecne the view when using "this," you must have a helper method
- To remove the current rightPanel and add a new view for a use case, you simply execute the following code in the action listener for the button which switches to that new view:
```java
main.remove(rightPanel);
main.add(YOUR_NEW_VIEW_HERE, BorderLayout.CENTER);
updateView();
```
- Note that one of the helper methods is a updateView() method, which is what updates the UI to reflect the changes made to the rightPanel (you must do this after the steps described above). Also, when switching back to the original dashboard, you must reuse the dashboardViewModel.firePropertyChanged() method which completes this task for you rather than duplicating the logic and implementing this in the logic of another use case which would potentially break Clean Architecture

**Review by hadi-naqvi — COMMENTED** [review]
Coding Style:
The spacing/indentation conventions of all the newly implemented code is consistent with the rest of the codebase. Additionally, the appropriate selection between public and private methods/attributes is consistent with the design choices made in other parts of the program. Furthermore, the variable naming conventions are also consistent with the other parts of the program, which includes the naming conventions of JComponents in our Swing UIs which was previously seen to be inconsistent in others parts of our program such as the SetupAuth and Authentication views. While those views need their variable names to be changed to be consistent, the variable names for this use case do not need to change, as they already match the majority of our codebase's variable naming conventions.

Correctness:
The purpose of this use case is to allow the user to securely store new login data in their password manager. This login data includes the title of their login data, username, password, 2FA secret key, url, iconURL (the icon of the service/website the account is for), date, and other notes related to the login information. This use case allows the user to do exactly that, while accounting for any potential errors. In the success case, a new account is created and added to the database, and the dashboard is updated to reflect the changes (i.e. adds the account to the JTable of accounts). Likewise in the fail case, if the user already has an account with the same title and username, it shows a popup indicating that they are attempting to create a duplicate account and the account is not created. Moreover, the implementaiton of this use case requires security, and it correctly encrypts the user's login data before storing it in the database correctly, and also empties the textfields and updates the state to be empty in the create account UI after account creation is complete as needed. This is important for correctness of security, because otherwise sensitive data (the login details) could potentially be accessed by foreign users.

This pull request also modifies the DisplayDash use case which is responsible for updating the dashboard's table of accounts with the most up to date list of accounts. It was modified so that the table is first emptied, and then the accounts are added back in to reflect the change. Previously the table wasn't emptied and so there would have been duplicate accounts listed in the table. Therefore since it was modified as required, this part of the use case's implementation is also correct and works as needed.

Presence of tests and Test coverage:
This pull request features no new tests or test coverage, however this is fine as test coverage is expected to be implemented shortly in another branch/pull request. As of right now, the use case's expected functionality has sufficiently been tested and shown to be correct, so this pull request can be merged into the main branch.

Adherence to Clean Architecture + SOLID:
All the application and business rules are segregated as needed. This use case follows the chain of going from the view, to the controller, to the use case interactor, to the presenter, which updates the view model, which then updates the view. This chain of events in which the use case occurs follows the Clean Architecture engine and design patterns that the rest of the use cases in the program have been using, and is not only consistent, but correctly applies Clean Architecture. Additionally, the dependency-inversion principle is not broken in this use case as all necessary components such as DAOs are being injected from the outer layers into the inner layers which require them. For example, in the app package which contains the use case factories, the interactor is instantiated with the DAO since it requires it, however the interactor itself does not have an inverse dependency on any of the outer layers. There is also a DataAccessInterface for this use case which dictates which methods the DAO must implement for the use case.

The SOLID principles in this use case have not been violated and have been followed well. A new design choice that was made in this pull request was to create separate views/forms using the Swing UI Designer in IntelliJ for the views which are going to be displayed in the right panel of the dashboard. Previously, our program was going to simply have all the views in the dashboard view. If we had done so, it would have been a violation of the single responsibility principle since the dashboard view would be responsible for multiple different views, and would have more than one purpose/reason to change which could introduce bugs in the future. As such, the current approach is ideal and follows the single responsibility principle well as it segregates different responsibilities/views into different files, and then instantiates the different views inside the dashboard view as needed.

Javadocs:
While most public methods are documented well with javadocs, there are some methods which are left without a Javadoc still, such as the switchView() method which includes the interface CreateAccountOutputBoundary, the controller, the use case interactor, and presenter for the use case (all of which have the method). Otherwise, all other methods have sufficient and well detailed documentation describing method purposes, parameters, and return types.

Clarification of any accompanying information in the pull request:
The accompanying information in this pull request touches upon most of the bug fixes, design changes, and new additions in this pull request. However there are some changes which are incorrectly explained or could be better explained for other readers and I will clarify them below:

The encrypt() and decrypt() methods in the DAO responsible for encrypting and decrypting our login data were updating to handle the edge case of the data being encrypted/decrypted being an empty string (i.e. when the user creates an account, they do not fill in one of the fields, which is something they are allowed to do as not all fields are necessarily required and that part is left up to the user). The methods now check if the data being encrypted/decrypyted is an empty string, and simply return an empty string back as nothing else needs to be done. Previously, it would attempt to run the encryption/decryption algorithm and result in errors/not work correctly. Note that although the pull request mentions whitespaces, there are no whitespaces in our program anywhere, just empty strings.
In the future, whenever the right panel view is changed to display another view, its rightPanelView name attribute must be updated accordingly (initially on program startup it is null which is fine). When exiting from the view of another use case with a cancel/close button, you will need to update that attribute name with "dashboard" and then trigger "dashboardViewModel.firePropertyChanged()" in order to switch the right panel view back to the original dashboard. For further information, take a look into the code to understand how this process works and the design choices made while implementing the logic for this.
The DashboardView file now has helper methods for operations requiring the "this" keyword inside of action listeners. This was necessary because inside the frame/context of an action listener, "this" refers to the ActionListener object and not the view itself. Therefore to referecne the view when using "this," you must have a helper method
To remove the current rightPanel and add a new view for a use case, you simply execute the following code in the action listener for the button which switches to that new view:
```java
main.remove(rightPanel);
main.add(YOUR_NEW_VIEW_HERE, BorderLayout.CENTER);
updateView();
```
Note that one of the helper methods is a updateView() method, which is what updates the UI to reflect the changes made to the rightPanel (you must do this after the steps described above). Also, when switching back to the original dashboard, you must reuse the dashboardViewModel.firePropertyChanged() method which completes this task for you rather than duplicating the logic and implementing this in the logic of another use case which would potentially break Clean Architecture

File: src/use_case/CreateAccount/CreateAccountOutputBoundary.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+     */
+    void prepareSuccessView(CreateAccountOutputData createAccountOutputData);
+
+    /**
+     * Method which updates the view manager model to display the failed CreateAccount screen
+     * @param error The error message which explains why CreateAccount failed
+     */
+    void prepareFailView(String error);
+
+    void switchView();
```
Comment (hadi-naqvi) [inline src/use_case/CreateAccount/CreateAccountOutputBoundary.java:16]: Missing Javadoc

File: src/use_case/CreateAccount/CreateAccountInteractor.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        for (AccountInfo account : userDataAccessObject.getAccounts()) {
+            if (account.getTitle().equals(title) && account.getUsername().equals(username)) {
+                return true; // Duplicate title and username combination found
+            }
+        }
+
+        return false; // No duplicate found
+    }
+
+    public void switchView(){
```
Comment (hadi-naqvi) [inline src/use_case/CreateAccount/CreateAccountInteractor.java:57]: Missing Javadoc

File: src/interface_adapter/CreateAccount/CreateAccountController.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+     */
+    public void execute(String title, String username, String password, String secretKey, String url,
+                        String iconURL, LocalDateTime date, String notes) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
+        CreateAccountInputData createAccountInputData = new CreateAccountInputData(title, username, password,
+                secretKey, url, iconURL, formatter.format(date), notes);
+        createAccountInteractor.execute(createAccountInputData);
+    }
+
+    public void switchView(){
```
Comment (hadi-naqvi) [inline src/interface_adapter/CreateAccount/CreateAccountController.java:32]: Missing Javadoc

File: src/interface_adapter/CreateAccount/CreateAccountPresenter.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+     * @param error The error message which explains why creating an account failed
+     */
+    @Override
+    public void prepareFailView(String error) {
+        CreateAccountState createAccountState = createAccountViewModel.getState();
+        createAccountState.setUsernameError(error);
+        this.createAccountViewModel.firePropertyChanged();
+    }
+
+    public void switchView(){
```
Comment (hadi-naqvi) [inline src/interface_adapter/CreateAccount/CreateAccountPresenter.java:52]: Missing Javadoc

File: src/use_case/CreateAccount/CreateAccountInputBoundary.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package use_case.CreateAccount;
+
+public interface CreateAccountInputBoundary {
+    /**
+     * Method which contains the logic for the CreateAccount use case (interactor) which is triggered to complete use case
+     * @param createAccountInputData The input data for the use case interactor
+     */
+    void execute(CreateAccountInputData createAccountInputData);
+
+    void switchView();
```
Comment (hadi-naqvi) [inline src/use_case/CreateAccount/CreateAccountInputBoundary.java:10]: Missing Javadoc

File: src/use_case/CreateAccount/CreateAccountOutputData.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    /**
+     * Constructor method for the CreateAccount output data
+     * @param useCaseFailed Whether the use case failed or not (i.e. failed creating account)
+     */
+    public CreateAccountOutputData(boolean useCaseFailed, List<AccountInfo> accounts) {
+        this.useCaseFailed = useCaseFailed;
+        this.accounts = accounts;
+    }
+
+    public List<AccountInfo> getAccounts(){
```
Comment (hadi-naqvi) [inline src/use_case/CreateAccount/CreateAccountOutputData.java:21]: Missing Javadoc
